### PR TITLE
Code review/3 atari

### PR DESCRIPTION
**Pull request by Hyeokreal**
코드리뷰 부탁드립니다!

### CODE REVIEW

File: 3-atari/1-breakout-dqn/test.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+        self.model = DQN(action_size, state_size)
+        self.model.load_weights(model_path)
+
+    def get_action(self, history):
+        history = np.float32(history / 255.0)
+        if np.random.rand() <= self.epsilon:
+            return random.randrange(self.action_size)
+        else:
+            q_value = self.model.predict(history)
```
Comment (dnddnjs) [inline 3-atari/1-breakout-dqn/test.py]: 저는 predict가 없어도 작동이 돼서 predict를 안 쓰고 있어요. 다음과 같은 식으로요

`q_value = self.model(history)`

File: 3-atari/1-breakout-dqn/train.py [hunk 2]
```diff
@@ -28,14 +24,12 @@ def __init__(self, action_size, state_size):
         self.fc_out = Dense(action_size)
```
Comment (dnddnjs) [inline 3-atari/1-breakout-dqn/train.py:25]: 마지막 출력층에 initializer 넣었다고 하지 않았어요??

File: 3-atari/1-breakout-dqn/train.py [hunk 3]
```diff
@@ -101,6 +97,7 @@ def train_model(self):
         dones = np.array([sample[4] for sample in mini_batch])
 
         model_params = self.model.trainable_variables
+
```
Comment (dnddnjs) [inline 3-atari/1-breakout-dqn/train.py]: 공백 빼기?

File: 3-atari/1-breakout-dqn/train.py [hunk 4]
```diff
@@ -101,6 +97,7 @@ def train_model(self):
...
         with tf.GradientTape() as tape:
             tape.watch(model_params)
```
Comment (dnddnjs) [inline 3-atari/1-breakout-dqn/train.py]: 다음 주석 추가

`현재 상태에 대한 모델의 큐함수`

**Review by dnddnjs — COMMENTED** [review]
cartpole dqn 코드에 있던 주석들은 겹치는 코드가 있으면 넣어주면 좋을 것 같아요

File: 3-atari/1-breakout-dqn/train.py [hunk 5]
```diff
@@ -47,36 +41,38 @@ def __init__(self, action_size, state_size=(84, 84, 4)):
...
+        self.writer = tf.summary.create_file_writer('summary/breakout_dqn')
 
-        self.model_path = os.path.join(os.getcwd(), 'save_model')
+        self.model_path = os.path.join(os.getcwd(), 'save_model', 'model')
         os.makedirs(self.model_path, exist_ok=True)
 
     def update_target_model(self):
         self.target_model.set_weights(self.model.get_weights())
 
+    def huber_loss(self, x):
```
Comment (dnddnjs) [inline 3-atari/1-breakout-dqn/train.py]: 이 함수는 사용하지 않는 것 같은데요??

**Review by dnddnjs — COMMENTED** [review]
하나 더 추가했어요

File: 3-atari/1-breakout-a3c/test.py [hunk 6]
```diff
@@ -2,50 +2,113 @@
 import gym
 import time
 import threading
```
Comment (dnddnjs) [inline 3-atari/1-breakout-a3c/test.py:4]: threading 패키지는 사용 안하는 거죠?

File: 3-atari/1-breakout-a3c/train.py [hunk 7]
```diff
@@ -58,25 +54,20 @@ def __init__(self, action_size):
...
         self.no_op_steps = 30
-        self.lr = 2.5e-4
+        self.lr = 1e-4
         # 쓰레드의 갯수
         self.threads = 16
 
         # 글로벌 인공신경망 생성
-        self.global_model = ActorCritic(self.action_size)
+        self.global_model = ActorCritic(self.action_size, self.state_size)
         self.global_model(tf.convert_to_tensor(np.random.random((1, *self.state_size)), dtype=tf.float32))
```
Comment (dnddnjs) [inline 3-atari/1-breakout-a3c/train.py:63]: 이거는 왜 하는 건가요? 주석으로 설명이 있어야 할 것 같아요.

File: 3-atari/1-breakout-a3c/train.py [hunk 8]
```diff
@@ -9,47 +9,43 @@
 from skimage.color import rgb2gray
 from skimage.transform import resize
 
-from tensorflow.compat.v1 import nn, train
+from tensorflow.keras.initializers import RandomUniform
+from tensorflow.compat.v1.train import RMSPropOptimizer, AdamOptimizer
```
Comment (dnddnjs) [inline 3-atari/1-breakout-a3c/train.py:13]: RMSPropOptimizer는 사용 안하는 패키지인 것 같네용

File: 3-atari/1-breakout-a3c/train.py [hunk 9]
```diff
@@ -259,11 +159,11 @@ def compute_loss(self, done):
...
         advantages = discounted_prediction - values
-        value_loss = tf.square(advantages)
+        value_loss = 0.5 * tf.reduce_sum(tf.square(advantages))
 
         # 정책 신경망 업데이트
         action = tf.convert_to_tensor(self.actions, dtype=tf.float32)
-        policy_prob = nn.softmax(policy)
+        policy_prob = tf.nn.softmax(policy)
         action_prob = tf.reduce_sum(action * policy_prob, axis=1, keepdims=True)
         cross_entropy = tf.math.log(action_prob + 1e-10) * tf.stop_gradient(advantages)
```
Comment (dnddnjs) [inline 3-atari/1-breakout-a3c/train.py:168]: 이런 식으로 바꾸면 어때요? 

`cross_entropy = - tf.math.log(action_prob + 1e-10)`
`policy_loss = tf.reduce_sume(cross_entropy * tf.stop_gradient(advantages))`

Comment (dnddnjs) [reply 3-atari/1-breakout-a3c/train.py:168]: cross_entropy 라는 의미가 좀 더 명확하게 되도록? 169, 176번째 줄도 변경하구요

File: 3-atari/1-breakout-a3c/train.py [hunk 10]
```diff
@@ -273,32 +173,109 @@ def compute_loss(self, done):
 
         policy_loss = cross_entropy + 0.01 * entropy
```
Comment (dnddnjs) [inline 3-atari/1-breakout-a3c/train.py:174]: policy_loss = policy_loss + 0.01 * entropy

**Review by dnddnjs — COMMENTED** [review]
여기도 전반적으로 주석이 잘 안 달려있는데 필요한 부분들은 달아주면 좋을 것 같아요. cartpole actor critic 참고해서?

File: 3-atari/1-breakout-dqn/train.py (hunk 2 above)
Comment (Hyeokreal) [reply 3-atari/1-breakout-dqn/train.py:25]: 아 이건 Adam 적용하고 나서부터는 빼는게 더 잘되더라구요!

File: 3-atari/1-breakout-dqn/test.py (hunk 1 above)
Comment (Hyeokreal) [reply 3-atari/1-breakout-dqn/test.py]: 넵 수정하겠습니다.

File: 3-atari/1-breakout-dqn/train.py (hunk 5 above)
Comment (Hyeokreal) [reply 3-atari/1-breakout-dqn/train.py]: 네 삭제 했습니다!

File: 3-atari/1-breakout-a3c/test.py (hunk 6 above)
Comment (Hyeokreal) [reply 3-atari/1-breakout-a3c/test.py:4]: 네네 테스트에서는 time 하고 os 패키지도 사용하고 있지 않아서 삭제했습니다!

File: 3-atari/1-breakout-a3c/train.py (hunk 7 above)
Comment (Hyeokreal) [reply 3-atari/1-breakout-a3c/train.py:63]: 로컬신경망으로 그레디언트를 계산해서 글로벌 신경망에 업데이트하기 때문에 글로벌 신경망에는 실질적으로 input 이 한번도 주어지지 않아서 저렇게 초기화를 해주지 않으면 오류가 납니다. (input 을 한번이라도 흘려줘야 graph shape 이 정의됩니다.)  그런데 저렇게 초기화 하는 방법은 너무 지저분해보여서 아래처럼  바꾸었습니다!
`self.global_model.build(tf.TensorShape((None, *self.state_size)))`

File: 3-atari/1-breakout-a3c/train.py (hunk 9 above)
Comment (Hyeokreal) [reply 3-atari/1-breakout-a3c/train.py:168]: 음.. cross entropy 가 - \sum p(x)log q(x) 이니까 다음 같이 변경하는건 어떤가요?

`        action = tf.convert_to_tensor(self.actions, dtype=tf.float32)`
`        policy_prob = tf.nn.softmax(policy)`
`        action_prob = tf.reduce_sum(action * policy_prob, axis=1, keepdims=True)`
`        cross_entropy = -tf.math.log(action_prob + 1e-10) * tf.stop_gradient(advantages)`
`        policy_loss = tf.reduce_sum(cross_entropy)`

File: 3-atari/2-breakout-a3c/train.py [hunk 11]
```diff
@@ -173,18 +183,18 @@ def compute_loss(self, done):
         action = tf.convert_to_tensor(self.actions, dtype=tf.float32)
         policy_prob = tf.nn.softmax(policy)
         action_prob = tf.reduce_sum(action * policy_prob, axis=1, keepdims=True)
-        cross_entropy = tf.math.log(action_prob + 1e-10) * tf.stop_gradient(advantages)
-        cross_entropy = -tf.reduce_sum(cross_entropy)
+        cross_entropy = -tf.math.log(action_prob + 1e-10) * tf.stop_gradient(advantages)
+        policy_loss = tf.reduce_sum(cross_entropy)
```
Comment (Hyeokreal) [inline 3-atari/2-breakout-a3c/train.py]: 이 부분만 검토 부탁드려요!! @dnddnjs

Comment (dnddnjs) [reply 3-atari/2-breakout-a3c/train.py]: 이렇게 하는 게 어떨까요?
```
cross_entropy = - tf.math.log(action_prob + 1e-10)
advantages = tf.stop_gradient(advantages)
actor_loss = tf.reduce_mean(cross_entropy * advantages)
```
